gpflow/actions.py

```diff
@@ -0,0 +1,386 @@
+        self._actions = actions
+    
+    def run(self, context: ActionContext):
+        """
+        Run executes list of cached actions in the for loop.
+        """
+        for action in self._actions:
+            action(context)
+
+
+class Condition(Action):
+    """
+    Condition actions is a branching mechanism.
+    Depending on boolean output of callable conditional function, either
+    true or false passed action is executed.
+
+        :param condition_fn: Boolean function with single input, current context.
+        :param action_true: True action, which is executed when output of
+            conditional function returns true.
+        :param action_false: False action, which is executed when output of
+            conditional function returns false.
+    """
+    def __init__(self,
+            condition_fn: Callable[[ActionContext], bool],
+            action_true: Action,
+            action_false: Optional[Action] = None) -> None:
+        self.condition_fn = condition_fn
+        self.action_false = action_false
+        self.action_true = action_true
+
+    def run(self, context: ActionContext):
+        cond = self.condition_fn(context)
+        if cond:
+            return self.action_true(context)
+        elif self.action_false is None:
+            return None
+        return self.action_false(context)
+
+
+class Optimization(Action):
+    """
+    Optimization is an action wrapper for GPflow optimizers.
+    It contains an optimizer, a model and TensorFlow minimization tensor
+    generated by the optimizer and model.
+    This action may be used as single optimization step.
+    """
+
+    def with_optimizer(self, optimizer) -> 'Optimization':
+        """
+        Replace optimizer.
+
+            :param optimizer: GPflow optimizer.
+            :return: Optimization instance self reference.
+        """
+        self._optimizer = optimizer
+        return self
+
+
+    def with_model(self, model: Model) -> 'Optimization':
+        """
+        Replace model.
+
+            :param model: GPflow model.
+            :return: Optimization instance self reference.
+        """
+        self._model = model
+        return self
+    
+    def with_optimizer_tensor(self, tensor: Union[tf.Tensor, tf.Operation]) -> 'Optimization':
+        """
+        Replace optimizer tensor.
+
+            :param model: Tensorflow tensor.
+            :return: Optimization instance self reference.
+        """
+        self._optimizer_tensor = tensor
+        return self
+    
+    def with_run_kwargs(self, **kwargs: Dict[str, Any]) -> 'Optimization':
+        """
+        Replace Tensorflow session run kwargs.
+        Check Tensorflow session run [documentation](https://www.tensorflow.org/api_docs/python/tf/Session).
+
+            :param kwargs: Dictionary of tensors as keys and numpy arrays or
+                primitive python types as values.
+            :return: Optimization instance self reference.
+        """
```

Comment:
I guess this should only be used through the public method run_kwargs(),
Currently there is no test that exercises this method?

Reply:
in fact, run_kwargs does not run anything, it is run's kwargs for tensorflow session or did I misunderstand you?

Reply:
I struggled to locate a place where the saved _run_args are used, but actually I now see it: it's obviously passed systematically to run